Add transmission mode sample analysis functions

easyXAFS opened the pull request:
I considered a while breaking these functions into smaller pieces, but decided to just have all of the calculations done at once and then the user can pull what parts they want from the results. Happy to reorganize based on feedback.

Tried to stick to a consistent style guide and match docstring formats to the existing package. 

Realizing as I'm writing this that I might want a little more explanation in the docstring about how the absorbance steps for each element are calculated. This is done by performing a polynomial fit to the pre-edge absorption (from -200 to -50 eV, hard-coded) and extrapolating that to the post-edge, then comparing against the actual computed post-edge absorption.

I thought about putting `formula_to_mass_fracs` and `mass_fracs_to_formula` in the `chemparser.py` module, but for now put everything in `xray.py`. 

For comparison to Hephaestus, I ran a few examples and uses this test code to validate:
```python
def test_against_hephaestus():
    results = analyze_transmission_sample_from_formula(
        formula='Fe2O3',
        energy=7162,
        absorp_total=1,
        area=1,
        density=5.24
    )
    assert abs(results['mass_total(mg)'] - 3.506) < 1e-3, f"Fe2O3 mass at 7162 eV from Hephaestus should be 3.506 mg, calculated {results['mass_total(mg)']} instead."
    assert abs(results['absorption_length(um)'] - 6.7) < 1e-1, f"Fe2O3 absorption length at 7162 eV from Hephaestus should be 6.7um, calculated {results['absorption_length(um)']} instead."

    results = analyze_transmission_sample_from_formula(
        formula='Cu',
        energy=9029,
        absorp_total=1,
        area=1,
        density=8.94,
    )
    assert abs(results['mass_total(mg)'] - 3.640) < 1e-3, f"Cu mass at 9029 eV from Hephaestus should be 3.640 mg, calculated {results['mass_total(mg)']} instead."
    assert abs(results['absorption_length(um)'] - 4.1) < 1e-1, f"Cu absorption length at 9029 eV from Hephaestus should be 4.1um, calculated {results['absorption_length(um)']} instead."
```
![hephaestus example](https://user-images.githubusercontent.com/59660036/118331999-42a7a100-b4be-11eb-8ef4-7c1c0a48a123.png)

For comparison to XAFSmass, here is a quick validation:
```python
analyze_transmission_sample_from_mass_fracs(
        mass_fracs={'Fe': 0.05, 'SiO2': -1},
        energy=xraydb.xray_edge('Fe', 'K').energy + 50,
        area=1.33,
        density=2.65
    )

{'absorbance_steps': {'Fe': 0.6692395733146204,
                      'O': 3.386553723091669e-07,
                      'Si': 1.297403071978392e-06},
 'absorp_total': 2.6,
 'absorption_length(um)': 55.71934579361293,
 'area(cm^2)': 1.33,
 'density(g/cc)': 2.65,
 'energy(eV)': 7162.0,
 'mass_components(mg)': {'Fe': 2.552976845244654,
                         'O': 25.832813088371587,
                         'Si': 22.673746971276834},
 'mass_fractions': {'Fe': 0.05,
                    'O': 0.5059351230797396,
                    'Si': 0.4440648769202603},
 'mass_total(mg)': 51.05953690489308,
 'thickness(mm)': 0.14487029906339363}
```
![xfasmass_example](https://user-images.githubusercontent.com/59660036/118332657-4982e380-b4bf-11eb-9d2a-61b4c563c3c4.png)

newville commented:
@easyXAFS That looks really good Will!  I like the agreement with Hephaestus and XAFSMass, but I would not get too hung up on slight differences (even to 10%) as the tabulated values do vary some.  

I think my main comments/questions would be:
   a) could the output be a named tuple, which would be more of a "guarantee" of what's included?
   b) could you combine the two functions by allowing the first argument to be either a string (formula) or dict (fracs), and maybe a keyword argument for whether those were "molar" or "mass" fractions?
  
   c) the function names are sort of long ;).  Combining the two optional input options might help, but maybe just call it `transmission_sample`?

Anyway, those are all minor points -- thanks!

easyXAFS commented:
@newville Awesome, thanks for the quick feedback. Yes since I knew they were using different tables, I was almost surprised with how close the results were (for Hephaestus they seem exact, since Hephaestus also uses the Elam tables).

Thanks for the comments:

> a) could the output be a named tuple, which would be more of a "guarantee" of what's included?

Ah yes that's a good idea. I think that fits better with the XrayDB style as well. I'll do this.

> b) could you combine the two functions by allowing the first argument to be either a string (formula) or dict (fracs), and maybe a keyword argument for whether those were "molar" or "mass" fractions?
> c) the function names are sort of long ;). Combining the two optional input options might help, but maybe just call it transmission_sample?

Totally agree function names are too long. As I was typing them I didn't like how verbose they were but couldn't quickly settle on an alternative. I like the keyword suggestion for b), I think I'll go with that and implement c).

easyXAFS commented:
@newville I am not well versed with PRs and updating PRs. Should I submit a new PR or am I missing a simple way to update the code in this PR?

easyXAFS commented:
@newville Okay pushed new changes addressing your comments, thanks for the useful suggestions. With the named tuple, I added the units as postfixes (e.g. `energy_eV`, `thickness_mm`). This contrasts with the other named tuples `DarwinWidth` and `fluxes` but there's a lot of different units floating around so thought it might be useful.

easyXAFS commented:
Also added top-level import which is nice, now it fits simply into the package:

```python
In [1]: import xraydb

In [2]: xraydb.transmission_sample('Fe2O3', 7162)
Out[2]: TransmissionSample(energy_eV=7162, absorp_total=2.6, mass_fractions={'Fe': 0.6994307614270416, 'O': 0.3005692385729583}, absorbance_steps={'Fe': 2.2227981005407176, 'O': 4.7769571901536886e-08}, area_cm2=None, mass_total_mg=None, mass_components_mg=None, density=None, thickness_mm=None, absorption_length_um=None)
```

easyXAFS commented:
One last change, decided to set `area=1` (cm^2) as default parameter. This is basically what Hephaestus does and it's nice to get some mass info out even if you forget to pick an area.

newville commented:
@easyXAFS Looks great -- thanks so much!  I'm going to merge this, but if you'd like to make more tweaks, feel free to keep going on this!
